pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,134 @@
+    from pylint.lint import PyLinter
+
+
+class ModifiedIterationChecker(checkers.BaseChecker):
+    """Checks for modified for loops iterations
+    Currently supports `for` loops for Sets, Dictionaries and Lists.
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+
+    name = "modified_iteration"
+
+    msgs = {
+        "W4701": (
+            "Iterated list '%s' is being modified inside loop body, consider iterating through a copy of it instead.",
+            "modified-iterating-list",
+            "Emitted when items are added or removed to a list being iterated through. "
+            "Doing so ccan result in unexpected behaviour, that is why it is preferred to use a copy of the list.",
+        ),
+        "E4702": (
+            "Iterated dict '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "modified-iterating-dict",
+            "Emitted when items are added or removed to a dict being iterated through. "
+            "Doing so raises a RuntimeError",
+        ),
+        "E4703": (
+            "Iterated set '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "modified-iterating-set",
+            "Emitted when items are added or removed to a set being iterated through. "
+            "Doing so raises a RuntimeError",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            msg_id = None
+            if self._modified_iterating_list_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-list"
+            elif self._modified_iterating_dict_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-dict"
+            elif self._modified_iterating_set_cond(body_node, iter_obj):
+                msg_id = "modified-iterating-set"
+            if msg_id is not None:
+                self.add_message(
+                    msg_id,
+                    node=node,
+                    args=(iter_obj.name,),
+                )
+                break  # since the msg is raised for the `for` node no further check is needed
+
+    @staticmethod
+    def _is_attribute_call_expr(body_node) -> bool:
+        return (
+            isinstance(body_node, astroid.Expr)
+            and isinstance(body_node.value, astroid.Call)
+            and isinstance(body_node.value.func, astroid.Attribute)
+        )
+
+    @classmethod
+    def _common_cond_list_set(cls, node, list_obj, infer_val) -> bool:
+        try:
+            return (infer_val == utils.safe_infer(list_obj)) and (
+                node.value.func.expr.name == list_obj.name
+            )
+        except AttributeError:
+            return False
+
+    @staticmethod
+    def _dict_node_cond(node) -> bool:
```

Comment:
```suggestion
    def _dict_node_cond(node: nodes.NodeNG) -> bool:
```

I think?
And could you add a one line docstring or make the method name a little more expressive?